File: backend/app/models/test_run.py
"""Models for configuration test runs."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class TestRunStatus(str, Enum):
    """Test run status."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class OutputFormat(str, Enum):
    """Output format for test run results."""

    PREVIEW = "preview"
    CSV = "csv"
    JSON = "json"


class TestRunOptions(BaseModel):
    """Options for running a configuration test."""

    entities: Optional[List[str]] = Field(None, description="List of entity names to test. If None, tests all entities")
    max_rows_per_entity: int = Field(100, description="Maximum number of rows to process per entity", ge=10, le=10000)
    output_format: OutputFormat = Field(OutputFormat.PREVIEW, description="Output format for results")
    validate_foreign_keys: bool = Field(True, description="Validate foreign key relationships")
    validate_constraints: bool = Field(True, description="Validate constraints")
    stop_on_error: bool = Field(False, description="Stop processing on first error")


class ValidationIssue(BaseModel):
    """A validation issue found during test run."""

    entity_name: str = Field(..., description="Entity where issue was found")
    severity: str = Field(..., description="Issue severity: error, warning, info")
    message: str = Field(..., description="Issue description")
    suggestion: Optional[str] = Field(None, description="Suggested fix")
    location: Optional[str] = Field(None, description="Where the issue occurred")


class EntityTestResult(BaseModel):
    """Test result for a single entity."""

    # See https://github.com/pylint-dev/pylint/issues/10087 for why pylint reports no-member for container fields
    entity_name: str = Field(..., description="Entity name")
    status: str = Field(..., description="Processing status: success, failed, skipped")
    rows_in: int = Field(..., description="Number of input rows")
    rows_out: int = Field(..., description="Number of output rows")
    execution_time_ms: int = Field(..., description="Execution time in milliseconds")
    error_message: Optional[str] = Field(None, description="Error message if failed")
    warnings: List[str] = Field(default_factory=list, description="Warning messages")
    validation_issues: List[ValidationIssue] = Field(default_factory=list, description="Validation issues found")
    preview_rows: Optional[List[Dict[str, Any]]] = Field(None, description="Preview of output rows (first 10)")


class TestProgress(BaseModel):
    """Progress information for a running test."""

    run_id: str = Field(..., description="Test run ID")
    status: TestRunStatus = Field(..., description="Current status")
    current_entity: Optional[str] = Field(None, description="Entity currently being processed")
    entities_completed: int = Field(..., description="Number of entities completed")
    entities_total: int = Field(..., description="Total number of entities to process")
    progress_percentage: float = Field(..., description="Progress percentage (0-100)")
    elapsed_time_ms: int = Field(..., description="Elapsed time in milliseconds")
    estimated_time_remaining_ms: Optional[int] = Field(None, description="Estimated time remaining in milliseconds")


class TestRunResult(BaseModel):
    """Complete test run result."""

    run_id: str = Field(..., description="Unique test run ID")
    config_name: str = Field(..., description="Configuration name")
    status: TestRunStatus = Field(..., description="Overall status")
    started_at: datetime = Field(..., description="Test run start time")
    completed_at: Optional[datetime] = Field(None, description="Test run completion time")
    total_time_ms: int = Field(..., description="Total execution time in milliseconds")
    entities_processed: List[EntityTestResult] = Field(default_factory=list, description="Results for each entity")
    entities_total: int = Field(0, description="Total number of entities to process")
    entities_succeeded: int = Field(0, description="Number of entities that succeeded")
    entities_failed: int = Field(0, description="Number of entities that failed")
    entities_skipped: int = Field(0, description="Number of entities that were skipped")
    validation_issues: List[ValidationIssue] = Field(default_factory=list, description="All validation issues found")
    error_message: Optional[str] = Field(None, description="Overall error message if failed")
    options: TestRunOptions = Field(..., description="Test run options used")
    current_entity: Optional[str] = Field(None, description="Entity currently being processed")
    entities_completed: int = Field(0, description="Number of entities completed")


class TestRunRequest(BaseModel):
    """Request to start a test run."""

    config_name: str = Field(..., description="Configuration name to test")
    options: TestRunOptions = Field(default_factory=TestRunOptions, description="Test run options")

File: frontend/src/types/testRun.ts
/**
 * Test Run Types
 *
 * TypeScript types matching the backend test run models.
 */

export enum TestRunStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum OutputFormat {
  PREVIEW = 'preview',
  CSV = 'csv',
  JSON = 'json',
}

export interface TestRunOptions {
  entities?: string[]
  max_rows_per_entity: number
  output_format: OutputFormat
  validate_foreign_keys: boolean
  validate_constraints: boolean
  stop_on_error: boolean
}

export interface ValidationIssue {
  severity: 'error' | 'warning' | 'info'
  entity: string
  issue_type: string
  message: string
  details?: Record<string, any>
}

export interface EntityTestResult {
  entity_name: string
  status: 'success' | 'failed' | 'skipped'
  rows_in: number
  rows_out: number
  execution_time_ms: number
  error_message: string | null
  warnings: string[]
  validation_issues: ValidationIssue[]
  preview_rows: Record<string, any>[] | null
}

export interface TestRunResult {
  run_id: string
  project_name: string
  status: TestRunStatus
  started_at: string
  completed_at: string | null
  total_time_ms: number
  entities_processed: EntityTestResult[]
  entities_total: number
  entities_succeeded: number
  entities_failed: number
  entities_skipped: number
  validation_issues: ValidationIssue[]
  error_message: string | null
  options: TestRunOptions
  current_entity: string | null
  entities_completed: number
}

export interface TestProgress {
  run_id: string
  status: TestRunStatus
  current_entity: string | null
  entities_total: number
  entities_completed: number
  entities_succeeded: number
  entities_failed: number
  entities_skipped: number
  started_at: string
  elapsed_time_ms: number
}

export interface TestRunRequest {
  project_name: string
  options: Partial<TestRunOptions>
}

// Default test run options
export const DEFAULT_TEST_RUN_OPTIONS: TestRunOptions = {
  max_rows_per_entity: 100,
  output_format: OutputFormat.PREVIEW,
  validate_foreign_keys: true,
  validate_constraints: true,
  stop_on_error: false,
}
